contributors/design-proposals/flexvolume-deployment.md

```diff
@@ -0,0 +1,140 @@
+# **Dynamic Flexvolume Plugin Discovery**
+
+## **Objective**
+
+Kubelet and controller-manager do not need to be restarted manually in order for new Flexvolume plugins to be recognized.
+
+## **Background**
+
+Beginning in version 1.8, the Kubernetes Storage SIG is putting a stop to accepting in-tree volume plugins and advises all storage providers to implement out-of-tree plugins. Currently, there are two recommended implementations: Container Storage Interface (CSI) and Flexvolume.
+
+[CSI](https://github.com/container-storage-interface/spec/blob/master/spec.md) provides a single interface that storage vendors can implement in order for their storage solutions to work across many different container orchestrators, and volume plugins are out-of-tree by design. This is a large effort, the full implementation of CSI is several quarters away, and there is a need for an immediate solution for storage vendors to continue adding volume plugins.
+
+[Flexvolume](https://github.com/kubernetes/community/blob/master/contributors/devel/flexvolume.md) is an in-tree plugin that has the ability to run any storage solution by executing volume commands against a user-provided driver on the Kubernetes host, and this currently exists today. However, the process of setting up Flexvolume is very manual, pushing it out of consideration for many users. Problems include having to copy the driver to a specific location in each node, manually restarting kubelet, and user's limited access to machines.
+
+An automated deployment technique is discussed in [Recommended Deployment Method](#recommended-driver-deployment-method). The crucial change required to enable this method is allowing kubelet and controller manager to dynamically discover plugin changes.
+
+
+## **Overview**
+
+When there is a modification of the driver directory, a notification is sent to the filesystem watch from kubelet or controller manager. When kubelet or controller-manager searches for plugins (such as when a volume needs to be mounted), if there is a signal from the watch, it probes the driver directory and loads currently installed drivers as volume plugins.
+
+
+## **Detailed Design**
```

Comment:
(a random point in the proposal to start a new separate thread so the discussion is organized)

# Mount propagation

I am preparing mount propagation for release 1.8 + possibility to run mount utilities (`mount.glusterfs`, `/usr/bin/rbd`, flex drivers, ...) in *pods* instead of running them on the *host*. Initially, I planned not to support flex volumes in the first release and get the implementation solid first, but now that there is someone else preparing dynamic probing of flex plugin it should not be hard to extend it.

See https://github.com/kubernetes/community/pull/589 for full details.

tl;dr, the proposal expects that pod with mount utilities will put a unix domain socket into `/var/lib/kubelet/plugin-sockets/<plugin name>`, e.g. `/var/lib/kubelet/plugin-sockets/kubernetes.io/glusterfs`. That can be easily extended to flex volumes:

* Flex volume author puts all necessary utilities, usual `/usr/libexec/kubernetes/kubelet-plugins/volume/exec/<vendor>~<driver>/<driver>` and a `volume-exec` daemon (shipped by Kubernetes) into an image.
* System admin runs the image as a DaemonSet with privileged pods that have shared mount-propagation on `/var/lib/kubelet` from the pod to the host. Pods in the daemon set run `volume-exec` daemon with proper parameters (namely the volume plugin name).
  * `volume-exec` daemon puts an unix socket into `/var/lib/kubelet/plugin-sockets/<vendor>~<driver>/<driver>` on the host.
* Probe in kubelet scans `plugin-sockets` and registers a new flex volume plugin for every socket it finds there. This is the same as discovery of new drivers as designed in this proposal.
* When kubelet wants to call the driver, it checks if `plugin-sockets/<vendor>~<driver>/<driver>` socket exists.
  * If if the socket does not exist, it uses plain old `os,.Exec` to execute the driver.
  * If the socket exists, it uses a gRPC API provided by `volume-exec` to execute stuff in the pod that runs `volume-exec`. As result, **`/usr/libexec/kubernetes/kubelet-plugins/volume/exec/<vendor>~<driver>/<driver>` is executed in the pod**. Due to shared mount propagation, the driver can run in the pod and kubelet will all mounts it creates.
  * (all this is already part of #589, there will be very little changes to flex volume implementation).

Installation of a driver is quite simple, no need to copy the driver from the pod to the host, creating one socket is enough. Upgrade of the daemon set is more complicated though, as any fuse daemons now run in the pod  and not on the host. If the pod is killed during the update all volumes that it served are unmounted. See #589 for details.

To sum it up, if the probe as suggested in this proposal is implemented, it should be fairly easy to extend it to scan also for the sockets. That's the only necessary change and we get completely containerized flex volumes. Question is if we then need to mess up with copying drivers from pods to the host and such. On the other way, #589 is just a proposal and it may be changed and who knows if it catches 1.8 and in what shape.